### ISSUE TITLE
Animated Tiles

### ISSUE THREAD

**Issue by carlospazuzu** [vnen/godot-tiled-importer#19]
Does this plugin support Animated Tiles? Couldn't make it work here..

**Comment by vnen** [comment]
No, because Godot does not have support for it.

**Comment by carlospazuzu** [comment]
Thanks for your answer! I will try to implement it myself!

**Comment by Danlos** [comment]
There is a workaround here: https://www.reddit.com/r/godot/comments/45rvwa/animated_tiles_in_tilemap_is_it_possible/
I am not quite familiar with this plugin internal functionality but if you want to, you can check the video posted there where it shows how he achieved that behavior you are requesting.

**Comment by vnen** [comment]
The workaround is based on a shader. It's not so hacky, but I'm unsure if it would fit this plugin. If someone manages to make it work in a neat way, I'll gladly accept a PR.

**Comment by HummusSamurai** [comment]
I took a look at the shader approach and it seems beyond hacky to me.

A more rational approach would be to attach a script to each imported TileMap node/layer and programmatically add the code we need.

The process function would iterate over the tiles and change a tile's frame every N seconds.
Unlike the shader approach, this wouldn't require splitting the tileset sheets.

From there it would be trivial to make each tile have it's own timing, which is not possible with the shader approach as long as the shader script is attached to the entire tilemap.

What do you think: neat, hacky, meh?

**Comment by TeddyDD** [comment]
I'm not sure but your approach may hit the performance. AFAIK modifying tile-map in runtime is slow.
What about importing animated tiles as AnimatedSprite instance?

**Comment by vnen** [comment]
If someone gets this working in a performant non-hacky way, I will merge it. Godot (hopefully) will eventually get animated tilemaps.

**Comment by HummusSamurai** [comment]
@TeddyDD But then how they be used in practice?

I'm trying to find a way to integrate them without making it hard to work with them, but for the life of me I can't think of another way.

**Comment by jonbonazza** [comment]
After discussion in the Godot Discord, the conclusion seems to be that this would be the only way to make it work currently. It's hacky, but could certainly be automated by the plugin until proper tile map animations are supported:
https://github.com/godotengine/godot/issues/9179

**Comment by vnen** [comment]
There is now an AnimatedTexture in Godot 3.1 (https://twitter.com/reduzio/status/1018869906840879104). If many people need I can also add a compatibility solution for Godot 3.0: use a regular ProxyTexture and add an AnimationPlayer in the scene to change the texture.

**Comment by tonyrh** [comment]
> There is now an AnimatedTexture in Godot 3.1 (https://twitter.com/reduzio/status/1018869906840879104).

Does that mean that animated tiles are supported by the importer now?

**Comment by vnen** [comment]
> Does that mean that animated tiles are supported by the importer now?

No, because 1) I didn't work on it yet (nor anyone else proposed an implementaion) and 2) AnimatedTexture is a feature of Godot 3.1, which is not released yet.

**Comment by McSora** [comment]
As I thought the new AnimatedTexture would bring salvation to my animated tile struggle i stand corrected... As for the AnimatedTexture does not work with AtlasTextures... which makes the whole thing a little bit more complicated.  
(https://twitter.com/reduzio/status/1018869909781180427)

I tried the whole day getting it to work and followed 3 approaches:

- check if tile in tileset node has an "animation" node and creating atlas textures using that information, which should work in theory but as i said before animatedtexture does not work with atlas textures...

- using the information of the "animation" node i created separate Image objects, crop'ed to tilesize, copied the texture information of the global_image to that separate image file and created an ImageTexture with this Image for every frame in the animation node. Again, this should work in theory but the ResourcesSaver crashes on .save of packed_scene when it tries to format the tileset (i guess..)

- in a last hope approach i did pretty much the same thing as in approach 2 only did i save the separated images as .png files and loaded the ImageTextures from these.... Fun fact it worked like a charm (on the second import) but it's just wrong as a single .tmx import created 1000 png files in my project directory in my case which than also had to be imported by godot first before they could be used by the second import of my .tmx file...

The whole thing would be pretty easy if AnimatedTexture would work with AtlasTexture but it doesn't...
Anyone got any other ideas? :sweat_smile:

**Comment by giulianob** [comment]
@McSora Your idea seems like it should work. Did you ever figure out why ResourceSaver crashed? Also, why in the last approach did you have so many PNG files if you were only saving them for animated tiles?

**Comment by danielscherzer** [comment]
I just tried to get animations to work. 
But my task was simpler because all my animated tiles were TileObjects. 

When they are imported by your plugin they are each created as a sprite. So I only had to adapt the region and vframes and hframes and add a new animation in an AnimationPlayer. All of which I guess could be done automatically by your plugin because it should get this information out of the tiled files. 

![screenshot](https://user-images.githubusercontent.com/8530595/65171951-58946280-da4c-11e9-8147-103ffc6aa6d2.PNG)

And if this approach works one could argue to convert all animated tiles into TileObjects on import and use the same approach for those. Is this something you would consider or am I thinking completely wrong?
